### PR TITLE
Fail fast if trying to set browser WebSocket headers

### PR DESCRIPTION
**Pull request by martinbonnin**
The WebSocket browser API unfortunately [doesn't support setting headers](https://stackoverflow.com/questions/4361173/http-headers-in-websockets-client-api/4361358#4361358) so it can be surprising. Fail fast in that case. Many thanks @sproctor for diving into this one.

### CODE REVIEW

File: build-logic/src/main/kotlin/Mpp.kt [hunk 1]
```diff
@@ -80,11 +83,22 @@ fun Project.configureMpp(
 
     if (enabledJs && withJs) {
       js(IR) {
-        nodejs {
-          testTask {
-            useMocha {
-              // Override default 2s timeout
-              timeout = "120s"
+        if (browserTest) {
+          println("browserTest")
```
Comment (BoD) [inline build-logic/src/main/kotlin/Mpp.kt:87]:
```suggestion
```

File: libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/ws/JsWebSocketEngine.kt [hunk 2]
```diff
@@ -83,16 +83,21 @@ actual class DefaultWebSocketEngine : WebSocketEngine {
   // so it can be accessed inside js("") function
   @Suppress("UNUSED_PARAMETER", "UnsafeCastFromDynamic", "UNUSED_VARIABLE", "LocalVariableName")
   private fun createWebSocket(urlString_capturingHack: String, headers: Headers): WebSocket {
-    val protocolHeaderNames = headers.names().filter { it.equals("sec-websocket-protocol", true) }
+    val (protocolHeaderNames, otherHeaderNames) = headers.names().partition { it.equals("sec-websocket-protocol", true) }
     val protocols = protocolHeaderNames.mapNotNull { headers.getAll(it) }.flatten().toTypedArray()
     return if (PlatformUtils.IS_NODE) {
+      println("using node")
```
Comment (BoD) [inline libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/ws/JsWebSocketEngine.kt:89]:
```suggestion
```

File: libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/ws/JsWebSocketEngine.kt [hunk 3]
```diff
@@ -83,16 +83,21 @@ actual class DefaultWebSocketEngine : WebSocketEngine {
...
       val ws_capturingHack = js("eval('require')('ws')")
       val headers_capturingHack: dynamic = object {}
       headers.forEach { name, values ->
         headers_capturingHack[name] = values.joinToString(",")
       }
       js("new ws_capturingHack(urlString_capturingHack, protocols, { headers: headers_capturingHack })")
     } else {
+      println("using browser")
```
Comment (BoD) [inline libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/ws/JsWebSocketEngine.kt:97]:
```suggestion
```

File: tests/kotlin-js-store-1.8/yarn.lock [hunk 4]
```diff
@@ -2,31 +2,312 @@
 # yarn lockfile v1
 
 
+"@colors/colors@1.5.0":
```
Comment (BoD) [inline tests/kotlin-js-store-1.8/yarn.lock:5]: Added by mistake I think :)

**Review by BoD — APPROVED** [review]
👍

File: build-logic/src/main/kotlin/Mpp.kt (hunk 1 above)
Comment (martinbonnin) [reply build-logic/src/main/kotlin/Mpp.kt]: 😬

File: libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/ws/JsWebSocketEngine.kt (hunk 2 above)
Comment (martinbonnin) [reply libraries/apollo-runtime/src/jsMain/kotlin/com/apollographql/apollo3/network/ws/JsWebSocketEngine.kt]: 🙏

File: tests/kotlin-js-store-1.8/yarn.lock (hunk 4 above)
Comment (martinbonnin) [reply tests/kotlin-js-store-1.8/yarn.lock:5]: The debug printf we mistakes but I think this one is on purpose. I added a new test framework (karma, for browser tests). I think it requires those dependencies. Let's see what CI has to say about it.

Comment (BoD) [reply tests/kotlin-js-store-1.8/yarn.lock:5]: Oh sorry I thought this came from testing the 1.8.20 branch. Nevermind!

**Pull request merged**